### DISCUSSION TITLE
Perf: Make window functions use joins where possible

### DISCUSSION THREAD

**Discussion by NickCrews** [duckdb/duckdb#6042] [General]
I'm not sure if I'm phrasing this problem statement correctly, but bear with me and it should be obvious. Thank you so much for your time!

I am trying to get the value count for a column, but broadcast back the original cardinality. The nice and clean SQL for this is
`SELECT count(col) OVER (PARTITION BY col)`, but this turns out to be very slow. See below:

```python
import duckdb

# create a table of 10 million random ints between 0 and 99
con = duckdb.connect(":memory:")
n = 10_000_000
con.execute(f"CREATE TABLE ints AS SELECT CAST(FLOOR(RANDOM() * 100) AS INT) as rand FROM range({n})")

# Method 1, with the cleanest SQL, but very slow.
# Takes 22 seconds
con.execute("SELECT rand, count(rand) OVER (PARTITION BY rand) as reps FROM ints").df()

# Method 2, messier SQL using groupby-then-join, but much faster:
# Takes .4 seconds
sql = """
WITH value_counts AS (
    SELECT rand, COUNT(*) AS reps FROM ints GROUP BY rand
)
SELECT ints.rand, reps FROM ints, value_counts WHERE ints.rand = value_counts.rand
"""
con.execute(sql).df()
```

I think it should be possible for duckdb to optimize anything that looks like `SELECT {AGG_FUNCTION}(col) OVER (PARTITION BY col)`.

Let me know next steps and if I can help further.

**Comment by hawkfish** [comment]
We actually added an optimization to the window code to check for constant aggregates like this for issue #6728. Just today we merged vectorization of the evaluation from issue #7852, which should make it even faster. I admit I haven't profiled this specific case, but I would be surprised if scanning a table twice and building a hash table was faster than scanning once while partitioning and sorting the partitions. But I'd be happy to be proved wrong!

**Reply by NickCrews** [reply]
Oh very nice! Thank you!

Will those changes be included in https://pypi.org/project/duckdb/0.8.2.dev157/?

I tried the following on that release and things look better, but still not equal:

```python
import timeit

import duckdb

# create a table of 10 million random ints between 0 and 99
con = duckdb.connect(":memory:")
n = 10_000_000
con.execute(
    f"CREATE TABLE ints AS SELECT CAST(FLOOR(RANDOM() * 100) AS INT) as rand FROM range({n})"
)

# Method 1, with the cleanest SQL, but very slow.
sql_slow = "SELECT rand, count(rand) OVER (PARTITION BY rand) as reps FROM ints"

# Method 2, messier SQL using groupby-then-join, but much faster:
sql_fast = """
WITH value_counts AS (
    SELECT rand, COUNT(*) AS reps FROM ints GROUP BY rand
)
SELECT ints.rand, reps FROM ints, value_counts WHERE ints.rand = value_counts.rand
"""


def bm(sql):
    return timeit.timeit(lambda: con.execute(sql).df(), number=10)


print("Slow SQL: ", bm(sql_slow))  # 11.2 seconds
print("Fast SQL: ", bm(sql_fast))  # 3.3 seconds
```

**Reply by hawkfish** [reply]
Thanks - this is interesting. Laurens is also making hash joins faster so its a moving target!

**Reply by NickCrews** [reply]
How do I look up if those commits were included in 0.8.2dev157? That was the latest release on pypi, but it was from the same day you replied that the fixes were merged, so I wasn't sure if I actually was seeing those commits

**Reply by NickCrews** [reply]
Hey @hawkfish I just tried this again in version 0.9.2 on a different (faster) machine and got

```
Slow SQL:  2.670828458001779
Fast SQL:  0.8347703750005167
```

If I bump up `n` to 100 million, I get a larger difference, 27.697967250001966 vs 6.990835209002398

So there is still that large speed difference. Is that to be expected after the improvements you mention above? Anything I can do to help the situation? Thank you!

**Reply by hawkfish** [reply]
No I think that makes sense. The windowed version hash partitions and sorts on the key, whereas the join version scans twice but only builds a hash table. One is `O(2 * N)` and the other is `O(N log N)`. But the difference appears to be about 3-4x in both cases and IIRC the original ratio was a lot worse.

Adding an optimiser rule for this might be worth doing, but it gets complicated because there may be other considerations. For example, if there was another window function that used the same partitioning but added ordering, then you can combine them into single window operator because the unordered one can ignore the order (this is an optimisation I plan to do soon). In that case the rewrite would be _slower_ because you have to do the partitioning anyway and you've just added another scan and hash table.

**Reply by NickCrews** [reply]
on duckdb 0.10.1, I now get `Slow SQL:  3.9379523750394583, Fast SQL:  3.0605128328315914`, so much better. Is there any hope of this continuing to improve, or are these as near-parity as I could expect?

**Reply by hawkfish** [reply]
I don't think there was anything done on the window operator since 0.9.2 that would have affected this query. The big changes were:

- Shared Window Partitions (the optimisation I mentioned on 2023-11-25 above, not used here)
- #4545: Windowed Distinct Aggregates (not used here either)

Other than that I can't think of anything unless some supporting library (like `PartitionedTupleData`) was improved and this flowed through.

Your query only partitions the data, so there is no sorting, just hash bin construction. So the physical plan difference is:

- Window: Single scan of the input, but materialise _all_ the rows in bins before computing the aggregate once per bin;
- Join: Scan the input twice, but only materialise the partition keys and the aggregate, before streaming a second scan through the hash table

The windowed plan will always require full materialisation of the partitioned input. This will typically be slower unless the table scan itself is expensive (e.g., reading from the network.) But if the scan is cheap and the bin count stays fixed, but the input size grows, the gap with the join will widen.  On the other hand, we could do the rewrite, but we would have to be able to cost the scan - and it would only apply when there is no `ORDER BY` clause in the window. 

Given this, leaving the optimisation up to the user is most likely. Plus, I don't think I would have used windowing to answer this question - the join approach seems much more natural. Windowing is complex and hard to make fast, and should only be used when nothing else can solve the problem.

**Reply by NickCrews** [reply]
Thank you for the very thorough explanation, it is really helpful for me to get a sense of the physical plans that get created so I can start to do that prediction myself.

Sounds good, I will probably stick with the join then. The reason I have been rooting for the window method is that I am using Ibis to construct the queries, not raw SQL. There, the API for the window approach is `def reps(col: Column) -> IntegerColumn`, but for the join approach it needs to be `def add_reps(t: Table, column_name: str) -> Table`. The first API ends up being MUCH more ergonomic, so I have been trying to use that, but the perf difference so far has forced me to use the second.

**Comment by NickCrews** [comment]
Some notes for myself testing different cardinalities. I tested on both v0.10.0 and v1.1.2, and there was no real difference:

## high cardinality (10k distinct values)

`CREATE OR REPLACE TABLE ints AS SELECT CAST(FLOOR(RANDOM() * 10_000) AS INT) as rand FROM range(100_000_000)`

~5.5sec vs ~3.5sec

## low cardinality (10 distinct values)

`CREATE OR REPLACE TABLE ints AS SELECT CAST(FLOOR(RANDOM() * 10) AS INT) as rand FROM range(100_000_000)`

~4.5sec vs ~2.5sec

As you would expect, the low-cardinality case is faster. But in both cases, the fast SQL takes about ~60% of the time

**Reply by hawkfish** [reply]
Be aware that your timings includes the conversion to a data frame, so it isn't exactly measuring the query times. Subtracting that overhead would make the join version proportionately faster(!)

**Comment by hawkfish** [comment]
[VLDB 2024 paper](https://www.vldb.org/pvldb/vol17/p2162-baca.pdf) on this subject.

**Comment by hawkfish** [comment]
@NickCrews I'm starting to look at this again bc there is now a community PR (#20081 ) doing something like this. That PR is rather limited, but I was thinking it could be generalised to what you suggest.

A few notes:

* You could do any number of aggregates here that follow the pattern
* We need `IS NOT DISTINCT` instead of equality to handle `NULL` partitioning elements
* Your `count` call could be changed to `count(*)` because there are no `NULL`s, which would be faster but require statistics

**Reply by hawkfish** [reply]
It's not any faster, but `count(*)` is logically an aggregate on a function of any column (col =>1 !!) so it could be included.

**Comment by hawkfish** [comment]
Just did a small test on 1.5 and the window times are now only 2.5x slower (0.108 vs 0.040), probably due to greater parallelism. (Note I was using the CLI with `.mode trash`)
